### PR TITLE
Migrate `shadow` custom properties from v11 to v12

### LINKED ISSUE #10641: Run `shadow` migration on next branch

### PR DESCRIPTION
**Pull request by laurkim**
### WHY are these changes introduced?

Resolves #10641.

Migrates `shadow` custom properties from v11 to v12.
Updates types on Box component `shadow` prop  and ShadowBevel component `boxShadow` prop.

### WHAT is this pull request doing?

To migrate `shadow` custom properties, the flow was to: 
- Handle automatic migrations
    ```sh
    # Space migration
    npx @shopify/polaris-migrator@0.0.0-snapshot-release-20230922144839 \
    v12-styles-replace-custom-property-shadow \
    'src/**/*.scss'
    # Stage all migrated files
    git add .
    # Format staged files only
    git diff --staged --name-only | xargs npx prettier --write
    # Commit automated migrations
    git commit -m "Migrate shadow custom properties from v11 to v12" 
    ```
- Handle manual migrations
    ```sh
    # Bring back the files with "polaris-migrator:" comments
    git stash pop
    ```
    - Search for "polaris-migrator:" comments and handle manual migrations
    - Search for and handle manual migrations (e.g. <Box shadow="md">)
    <br>

    ```sh
    # Stage all manually migrated files 
    git add .
    # Format staged files only
    git diff --staged --name-only | xargs npx prettier --write 
    #  Commit manual migrations
    git commit -m "Manually migrate shadow custom properties from v11 to v12"

    ```
- Check for errors after all custom property migrations are finished
    ```sh
    # Run stylelint against polaris-react
    npx stylelint 'polaris-react/src/**/*.scss'
    ```

#### Post-migration validation

After migrating use the following RegExp to check for any additional instances of `space` custom properties across all file types:

<details>
  <summary>[V12] Check RegExp for hardcoded <code>shadow</code> custom properties across all file types</summary>

```
(?:--p-shadow-inset-lg|--p-shadow-inset-md|--p-shadow-inset-sm|--p-shadow-none|--p-shadow-xs|--p-shadow-sm|--p-shadow-md|--p-shadow-lg|--p-shadow-xl|--p-shadow-2xl|--p-shadow-bevel-experimental|--p-shadow-card-sm-experimental|--p-shadow-card-md-experimental|--p-shadow-card-lg-experimental|--p-shadow-button-experimental|--p-shadow-button-hover-experimental|--p-shadow-button-disabled-experimental|--p-shadow-button-primary-strong-experimental|--p-shadow-button-primary-strong-inset-experimental|--p-shadow-button-primary-strong-hover-experimental|--p-shadow-border-inset-experimental|--p-shadow-button-primary-experimental|--p-shadow-button-primary-hover-experimental|--p-shadow-button-inset-experimental)(?![\w-])
```
</details>

<details>
  <summary>Check RegExp for outdated <code>&lt;Box shadow="..." /&gt;</code> props</summary>

```
<Box[^>\w](?:[^>]|\n)*?shadow
```
</details>

<details>
  <summary>Check RegExp for outdated <code>&lt;ShadowBevel boxShadow="..." /&gt;</code> props</summary>

```
<ShadowBevel[^>\w](?:[^>]|\n)*?boxShadow
```
</details>

#### Replacement maps

| Deprecated CSS Custom Property   | Replacement Value |
| ---------------- | ----------------- |
| `--p-shadow-inset-lg` | `--p-shadow-inset-200`    |
| `--p-shadow-inset-md` | `--p-shadow-inset-200`    |
| `--p-shadow-inset-sm` | `--p-shadow-inset-100`    |
| `--p-shadow-none` | `--p-shadow-0`    |
| `--p-shadow-xs` | `--p-shadow-100`    |
| `--p-shadow-sm` | `--p-shadow-200`    |
| `--p-shadow-md` | `--p-shadow-300`    |
| `--p-shadow-lg` | `--p-shadow-400`    |
| `--p-shadow-xl` | `--p-shadow-500`    |
| `--p-shadow-2xl` | `--p-shadow-600`    |
| `--p-shadow-bevel-experimental` | `--p-shadow-bevel-100`    |
| `--p-shadow-card-sm-experimental` | `--p-shadow-100`    |
| `--p-shadow-card-md-experimental` | `--p-shadow-200`    |
| `--p-shadow-card-lg-experimental` | `--p-shadow-300`    |
| `--p-shadow-button-experimental` | `--p-shadow-button`    |
| `--p-shadow-button-hover-experimental` | `--p-shadow-button-hover`    |
| `--p-shadow-button-disabled-experimental` | `inset 0 0 0 1px rgba(227, 227, 227, 1)`    |
| `--p-shadow-button-primary-strong-experimental` | `--p-shadow-button-primary`    |
| `--p-shadow-button-primary-strong-inset-experimental` | `--p-shadow-button-primary-inset`    |
| `--p-shadow-button-primary-strong-hover-experimental` | `--p-shadow-button-primary-hover`    |
| `--p-shadow-border-inset-experimental` | `--p-shadow-border-inset`    |

> [!IMPORTANT]  
> The following values were manually migrated based on context:

| Deprecated CSS Custom Property            | Replacement Value                  |
| ------------------------- | ------------------------ |  
| `--p-shadow-button-primary-experimental` | `--p-shadow-button-primary-critical`<br>`--p-shadow-button-primary-success`    |
| `--p-shadow-button-primary-hover-experimental` | `--p-shadow-button-primary-critical-hover`<br>`--p-shadow-button-primary-success-hover`    |
| `--p-shadow-button-inset-experimental` | `--p-shadow-button-primary-critical-inset`<br>`--p-shadow-button-primary-success-inset`    |

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris/blob/main/README.md#local-development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris/blob/main/.github/CONTRIBUTING.md#changelog)

<details>
<summary>Copy-paste this code in <code>playground/Playground.tsx</code>:</summary>

```jsx
import React from 'react';
import {Page} from '../src';

export function Playground() {
  return (
    <Page title="Playground">
      {/* Add the code you want to test in here */}
    </Page>
  );
}
```

</details>

### 🎩 checklist

- [ ] Tested on [mobile](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md#cross-browser-testing)
- [x] Tested on [multiple browsers](https://help.shopify.com/en/manual/shopify-admin/supported-browsers)
- [ ] Tested for [accessibility](https://github.com/Shopify/polaris/blob/main/documentation/Accessibility%20testing.md)
- [ ] Updated the component's `README.md` with documentation changes
- [ ] [Tophatted documentation](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting%20documentation.md) changes in the style guide

### CODE REVIEW

File: polaris-react/src/components/Button/Button.scss [hunk 1]
```diff
@@ -328,20 +328,13 @@
       top: 0.5px;
       bottom: 0.5px;
       border-radius: inherit;
-      // adjust shine for critical and success buttons
-      box-shadow: var(--p-shadow-button-primary-experimental);
```
Comment (laurkim) [inline polaris-react/src/components/Button/Button.scss:331]: Shadow for critical and success primary buttons were combined so I've removed them here to add to the individual classes instead with the token replacement value.

File: polaris-react/src/components/Button/Button.scss [hunk 2]
```diff
@@ -328,20 +328,13 @@
...
 
     &:hover {
       // stylelint-disable-next-line polaris/conventions/polaris/custom-property-allowed-list -- Button custom property
       --pc-button-text: var(--p-color-text-inverse);
 
-      // stylelint-disable-next-line selector-max-specificity -- set box shadow
-      &::before {
-        box-shadow: var(--p-shadow-button-primary-experimental);
-      }
-
```
Comment (laurkim) [inline polaris-react/src/components/Button/Button.scss:343]: Shadow for critical and success primary button hover states were combined so I've removed them here to add to the individual classes instead with the token replacement value.

File: polaris-react/src/components/Button/Button.scss [hunk 3]
```diff
@@ -351,7 +344,6 @@
     &:active {
       // stylelint-disable-next-line polaris/conventions/polaris/custom-property-allowed-list -- Button custom property
       --pc-button-text: var(--p-color-text-inverse-subdued);
-      box-shadow: var(--p-shadow-button-inset-experimental);
```
Comment (laurkim) [inline polaris-react/src/components/Button/Button.scss:353]: Shadow for critical and success primary button active states were combined so I've removed them here to add to the individual classes instead with the token replacement value.

File: polaris-tokens/src/themes/base/shadow.ts [hunk 4]
```diff
@@ -111,7 +111,7 @@ export const shadow: {
   },
   'shadow-button-primary': {
     value:
-      '0 1px 0 0 rgba(255, 255, 255, 0.24) inset, 1px 0 0 0 rgba(255, 255, 255, 0.20) inset, -1px 0 0 0 rgba(255, 255, 255, 0.20) inset, 0 -1px 0 0 #000 inset, 0 -1px 0 1px #1A1A1A',
+      '0px 2px 0px 0px rgba(255, 255, 255, 0.2) inset, 2px 0px 0px 0px rgba(255, 255, 255, 0.2) inset, -2px 0px 0px 0px rgba(255, 255, 255, 0.2) inset, 0px -1px 0px 1px #000 inset, 0px 1px 0px 0px #000 inset',
```
Comment (laurkim) [inline polaris-tokens/src/themes/base/shadow.ts:114]: Updated this token value after discussing with @bernardojoaogarcia and @sarahill when I noticed shadow on primary Button looked off.

Before with prior token value:
29-20-zs87f-wrdxd

**Review by lgriffee — APPROVED** [review]

File: polaris-tokens/src/themes/base/shadow.ts [hunk 5]
```diff
@@ -87,7 +87,7 @@ export const shadow: {
   },
   'shadow-bevel-100': {
     value:
-      '0 1px 0 0 rgba(208, 208, 208, 0.40) inset, 1px 0 0 0 #E3E3E3 inset, -1px 0 0 0 #E3E3E3 inset, 0 -1px 0 0 #D4D4D4 inset',
+      '1px 0px 0px 0px rgba(0, 0, 0, 0.13) inset, -1px 0px 0px 0px rgba(0, 0, 0, 0.13) inset, 0px -1px 0px 0px rgba(0, 0, 0, 0.17) inset, 0px 1px 0px 0px rgba(204, 204, 204, 0.5) inset',
   },
```
Comment (laurkim) [inline polaris-tokens/src/themes/base/shadow.ts:91]: Updated this token value after discussing with @sarahill and @bernardojoaogarcia when we noticed shadow bevel was a bit off with the previous `shadow-bevel-100` value.

**Review by bernardojoaogarcia — APPROVED** [review]

**Review by sarahill — APPROVED** [review]

**Pull request merged**